File: SPPE/src/models/layers/PRM.py
import torch.nn as nn
from .util_models import ConcatTable, CaddTable, Identity
import math
from opt import opt


class Residual(nn.Module):
    def __init__(self, numIn, numOut, inputResH, inputResW, stride=1,
                 net_type='preact', useConv=False, baseWidth=9, cardinality=4):
        super(Residual, self).__init__()

        self.con = ConcatTable([convBlock(numIn, numOut, inputResH,
                                          inputResW, net_type, baseWidth, cardinality, stride),
                                skipLayer(numIn, numOut, stride, useConv)])
        self.cadd = CaddTable(True)

    def forward(self, x):
        out = self.con(x)
        out = self.cadd(out)
        return out


def convBlock(numIn, numOut, inputResH, inputResW, net_type, baseWidth, cardinality, stride):
    numIn = int(numIn)
    numOut = int(numOut)

    addTable = ConcatTable()
    s_list = []
    if net_type != 'no_preact':
        s_list.append(nn.BatchNorm2d(numIn))
        s_list.append(nn.ReLU(True))

    conv1 = nn.Conv2d(numIn, numOut // 2, kernel_size=1)
    if opt.init:
        nn.init.xavier_normal(conv1.weight, gain=math.sqrt(1 / 2))
    s_list.append(conv1)

    s_list.append(nn.BatchNorm2d(numOut // 2))
    s_list.append(nn.ReLU(True))

    conv2 = nn.Conv2d(numOut // 2, numOut // 2,
                      kernel_size=3, stride=stride, padding=1)
    if opt.init:
        nn.init.xavier_normal(conv2.weight)
    s_list.append(conv2)

    s = nn.Sequential(*s_list)
    addTable.add(s)

    D = math.floor(numOut // baseWidth)
    C = cardinality
    s_list = []

    if net_type != 'no_preact':
        s_list.append(nn.BatchNorm2d(numIn))
        s_list.append(nn.ReLU(True))

    conv1 = nn.Conv2d(numIn, D, kernel_size=1, stride=stride)
    if opt.init:
        nn.init.xavier_normal(conv1.weight, gain=math.sqrt(1 / C))

    s_list.append(conv1)
    s_list.append(nn.BatchNorm2d(D))
    s_list.append(nn.ReLU(True))
    s_list.append(pyramid(D, C, inputResH, inputResW))
    s_list.append(nn.BatchNorm2d(D))
    s_list.append(nn.ReLU(True))

    a = nn.Conv2d(D, numOut // 2, kernel_size=1)
    a.nBranchIn = C
    if opt.init:
        nn.init.xavier_normal(a.weight, gain=math.sqrt(1 / C))
    s_list.append(a)

    s = nn.Sequential(*s_list)
    addTable.add(s)

    elewiswAdd = nn.Sequential(
        addTable,
        CaddTable(False)
    )
    conv2 = nn.Conv2d(numOut // 2, numOut, kernel_size=1)
    if opt.init:
        nn.init.xavier_normal(conv2.weight, gain=math.sqrt(1 / 2))
    model = nn.Sequential(
        elewiswAdd,
        nn.BatchNorm2d(numOut // 2),
        nn.ReLU(True),
        conv2
    )
    return model


def pyramid(D, C, inputResH, inputResW):
    pyraTable = ConcatTable()
    sc = math.pow(2, 1 / C)
    for i in range(C):
        scaled = 1 / math.pow(sc, i + 1)
        conv1 = nn.Conv2d(D, D, kernel_size=3, stride=1, padding=1)
        if opt.init:
            nn.init.xavier_normal(conv1.weight)
        s = nn.Sequential(
            nn.FractionalMaxPool2d(2, output_ratio=(scaled, scaled)),
            conv1,
            nn.UpsamplingBilinear2d(size=(int(inputResH), int(inputResW))))
        pyraTable.add(s)
    pyra = nn.Sequential(
        pyraTable,
        CaddTable(False)
    )
    return pyra


class skipLayer(nn.Module):
    def __init__(self, numIn, numOut, stride, useConv):
        super(skipLayer, self).__init__()
        self.identity = False

        if numIn == numOut and stride == 1 and not useConv:
            self.identity = True
        else:
            conv1 = nn.Conv2d(numIn, numOut, kernel_size=1, stride=stride)
            if opt.init:
                nn.init.xavier_normal(conv1.weight, gain=math.sqrt(1 / 2))
            self.m = nn.Sequential(
                nn.BatchNorm2d(numIn),
                nn.ReLU(True),
                conv1
            )

    def forward(self, x):
        if self.identity:
            return x
        else:
            return self.m(x)

File: train/src/models/layers/PRM.lua
-- Residual Pyramid with Pooling
local nnlib = require('cudnn')
local conv = nnlib.SpatialConvolution
local batchnorm = nnlib.SpatialBatchNormalization
local relu = nnlib.ReLU

-- Main convolutional block
local function convBlock(numIn,numOut,inputResH,inputResW,type,baseWidth, cardinality,stride)
    local model = nn.Sequential()
    local addTable = nn.ConcatTable()

    -- main branch
    local s = nn.Sequential()
    if type ~= 'no_preact' then
        s:add(batchnorm(numIn))
        s:add(relu(true))        
    end

    s:add(conv(numIn,numOut/2,1,1))
    s:add(batchnorm(numOut/2))
    s:add(relu(true))
    s:add(conv(numOut/2,numOut/2,3,3,stride,stride,1,1))
    addTable:add(s)

    -- Pyramid
    local function pyramid(D, C)
        local pyraTable = nn.ConcatTable()
        local sc = 2 ^(1/C);
        for i = 1, C do
            local scaled = 1/(sc^i)
            local s = nn.Sequential()
                :add(nn.SpatialFractionalMaxPooling(2, 2, scaled, scaled))
                :add(conv(D,D,3,3,1,1,1,1))
                :add(nn.SpatialUpSamplingBilinear({oheight=inputResH, owidth=inputResW}))
            pyraTable:add(s)
        end
        local pyra = nn.Sequential()
            :add(pyraTable)
            :add(nn.CAddTable(false))
        return pyra
    end


    local D = math.floor(numOut/baseWidth)
    local C = cardinality
    local s = nn.Sequential()

    if type ~= 'no_preact' then
        s:add(batchnorm(numIn))
        s:add(relu(true))        
    end
    local a = conv(D,numOut/2,1,1)
    a.nBranchIn = C
    s:add(conv(numIn,D,1,1,stride,stride))
    s:add(batchnorm(D))
    s:add(relu(true))
    s:add(pyramid(D, C))
    s:add(batchnorm(D))
    s:add(relu(true))
    -- s:add(conv(D,numOut/2,1,1))
    s:add(a)

    -- add together
    addTable:add(s)
    
    local elewiseAdd = nn.Sequential()
        :add(addTable)
        :add(nn.CAddTable(false))

    -- combine model
    return model:add(elewiseAdd)
        :add(batchnorm(numOut/2))
        :add(relu(true))
        :add(conv(numOut/2,numOut,1,1))
end

-- Skip layer
local function skipLayer(numIn,numOut,stride, useConv)
    if useConv then
        return nn.Sequential()
            :add(batchnorm(numIn))
            :add(relu(true))        
            :add(conv(numIn,numOut,1,1,stride,stride))
    end

    if numIn == numOut and stride == 1 then
        return nn.Identity()
    else
        return nn.Sequential()
            :add(batchnorm(numIn))
            :add(relu(true))        
            :add(conv(numIn,numOut,1,1,stride,stride))
    end
end

-- Residual block
--function Residual(numIn,numOut,stride,type,useConv,inputRes,baseWidth,cardinality)
--    local type = type or 'preact'
--    local baseWidth = baseWidth or 9
--    local cardinality = cardinality or 4
--    local useConv = useConv or false
--    local stride = stride or 1
--    return nn.Sequential()
--        :add(nn.ConcatTable()
--            :add(convBlock(numIn,numOut,inputRes,type,baseWidth,cardinality,stride))
--            :add(skipLayer(numIn,numOut,stride,useConv)))
--        :add(nn.CAddTable(true))
--end
function Residual(numIn,numOut,stride,type,useConv,inputResH,inputResW,baseWidth,cardinality)
    local type = type or 'preact'
    local baseWidth = baseWidth or 9
    local cardinality = cardinality or 4
    local useConv = useConv or false
    local stride = stride or 1
    return nn.Sequential()
        :add(nn.ConcatTable()
            :add(convBlock(numIn,numOut,inputResH,inputResW,type,baseWidth,cardinality,stride))
            :add(skipLayer(numIn,numOut,stride,useConv)))
        :add(nn.CAddTable(true))
end

return Residual
